File: mobile/android/app/src/main/java/com/theninjarpg/app/TNRWidgetSyncPlugin.java
package com.theninjarpg.app;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/** Android half of app/src/libs/native/widgetBridge.ts. */
@CapacitorPlugin(name = "TNRWidgetSync")
public class TNRWidgetSyncPlugin extends Plugin {

    @PluginMethod
    public void sync(PluginCall call) {
        String snapshot = call.getString("snapshot");
        if (snapshot == null) {
            call.reject("A snapshot is required");
            return;
        }
        TNRSnapshotStore.save(getContext(), snapshot);
        call.resolve();
    }

    @PluginMethod
    public void clear(PluginCall call) {
        TNRSnapshotStore.clear(getContext());
        call.resolve();
    }
}

File: mobile/ios/App/App/Plugins/TNRWidgetSyncPlugin.swift
import Capacitor
import Foundation
import WidgetKit

/// Writes the widget snapshot into the App Group and asks WidgetKit to redraw.
///
/// The web app calls this on every foreground, so the widgets stay right while the app is
/// closed and never have to make a network request of their own.
@objc(TNRWidgetSyncPlugin)
public class TNRWidgetSyncPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "TNRWidgetSyncPlugin"
    public let jsName = "TNRWidgetSync"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "sync", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise),
    ]

    @objc func sync(_ call: CAPPluginCall) {
        guard let snapshot = call.getString("snapshot") else {
            call.reject("A snapshot is required")
            return
        }
        TNRSnapshotStore.save(json: snapshot)
        reloadWidgets()
        call.resolve()
    }

    @objc func clear(_ call: CAPPluginCall) {
        TNRSnapshotStore.clear()
        reloadWidgets()
        call.resolve()
    }

    private func reloadWidgets() {
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
    }
}
